consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -918,26 +945,91 @@ impl<
 
         // Request verification
         debug!(?proposal, "requested proposal verification",);
-        let context = Context {
-            round: proposal.round,
-            leader: self
-                .scheme
-                .participants()
-                .key(leader)
-                .expect("leader not found")
-                .clone(),
-            parent: (proposal.parent, *parent_payload),
-        };
+        let context =
+            self.build_context(proposal.round, leader, (proposal.parent, *parent_payload));
         let proposal = proposal.clone();
         let payload = proposal.payload;
         let round = self.views.get_mut(&context.view()).unwrap();
         round.requested_proposal_verify = true;
-        Some((
-            context.clone(),
-            self.automaton.verify(context, payload).await,
-        ))
+        let receiver = self.automaton.verify(context.clone(), payload).await;
+        Some(Request(context, receiver))
+    }
+
+    /// Returns a request to certify a proposal if it should be requested.
+    ///
+    /// Returns `None` if the proposal should not be certified for any reason, for example if it
+    /// already has a certification request or does not have enough information to make the request.
+    async fn certify(
+        &mut self,
+        view: View,
+        resolver: &mut resolver::Mailbox<S, D>,
+    ) -> Option<Request<D, P, bool>> {
+        let round = self.views.get_mut(&view)?;
+
+        // Request certification only once
+        if round.certify_handle.is_some() || round.certified_proposal.is_some() {
+            return None;
+        }
+
+        // Request certification only if we have a notarization
+        round.notarization.as_ref()?;
+
+        // Get the proposal
+        let proposal = round
+            .proposal
+            .clone()
+            .expect("notarized proposal should have a proposal");
+
+        // Get the context for the proposal.
+        let parent_view = proposal.parent;
+
+        // If the leader or the parent for this view are not set, we need to fetch them.
+        let mut missing_notarizations = None;
+        let mut missing_nullifications = None;
+
+        // If the leader is not set, we need the seed for this round. If the parent view is not
+        // the previous view, then a nullification is missing.
+        let leader_opt = self.get_leader(self.view);
+        if leader_opt.is_none() {
+            let prev_view = view - 1;
+            if parent_view != prev_view {
+                missing_nullifications = Some(prev_view);
+            }
+        }
+
+        // Fetch the parent notarization if we don't have it. We need to guarantee the parent
+        // payload in order to certify this view.
+        let parent_payload_opt = self.is_notarized(parent_view).copied();
+        if parent_payload_opt.is_none() {
+            missing_notarizations = Some(parent_view);
```

Comment:
I guess we may still need to fetch this for the seed (even if we added parent digest to `Proposal`)?